### ISSUE TITLE
[lit-html] Attribute templates

### ISSUE THREAD

**Issue by justinfagnani** [lit/lit#1862]
We have several use cases for "attribute templates". These are templates that contain attribute-position bindings like attributes, properties, boolean attributes and other element bindings, but don't contain a tag name.

```
attr`a=${} .b=${} @c=${} ${}`
```

Attribute templates would be allowed in element expression position:

```
html`<div ${attr`...`}></div>`
```

While the inline case above isn't useful, attr templates are useful when you can apply them to an existing element or share them across many elements in different templates.

### Host properties:
```
class MyElement extends LitElement {
  renderHost() {
    return attr`selected=${this.selected} @click=${this._onClick}`;
  }
}
```

### Assigning properties to slotted children:
```
html`<slot ${eachAssigned((e) => attr`.prop=${value}`)}></slot>`
```

### Controllers returning a set of bindings to make on an element:

```
class RippleController {
  attach() {
    return attr`@click=${...} @mousedown=${...} @touchstart=${...}`
  }
}

html`
  <div ${this.rippleController.attach()}>
    ...
  </div>
`
```

**Comment by Vac1911** [comment]
> While the inline case above isn't useful, attr templates are useful when you can apply them to an existing element or share them across many elements in different templates.

I disagree, I think that inline case is also useful. For example it would allow you to use a single ternary operator where before you had to either use many ternaries or use 1 if statement but repeat all the child elements this:

```
return html`
    <div
        title="Card Option 2"
        value="2"
        ${this.active
            ? attr`class="bordered" styles="background-color: ${this.favColor}; font-weight: bold;" @click="${this.unsetActive}"`
            : attr`class="borderless" styles="background-color: ${getComplementaryColor(this.favColor)}" @click="${this.setActive}"` @mouseover="${this.showCardCaption}"`
        }
    >
        <div class="card-option-header">
           Choose Card Option 2!
        </div>
        <div class="card-option-body">
            <img />
            <div class="card-caption" ?visible="${this.cardCaptionIsVisible()}">
                More about this option
            </div>
        </div>
    </div>`
```

**Comment by Matsuuu** [comment]
I could be interested in tackling this one. 

I'll take it something similiar, like [ref](https://github.com/lit/lit/blob/main/packages/lit-html/src/directives/ref.ts) would be a good starting point to go from?

**Comment by Matsuuu** [comment]
![image](https://user-images.githubusercontent.com/16068444/118709810-21301780-b826-11eb-85f4-1a4320f18d27.png)

I've worked on this on a couple of nights and got some working PoC functioning today.

I approached this by doing some of the same sort of parsing the `html` -template literal does, and went from there. The attributes aren't added to the _parts list of the template, but I'm not quite sure if that's needed for this case. Would need more testing and fiddling around.

But if for example @justinfagnani you could check out if I'm at all going in the right direction, and we go from there?

My fork of the repo: https://github.com/Matsuuu/lit-html/blob/main/packages/lit-html/src/directives/attr.ts

**Comment by nmattia** [comment]
Hi! @Matsuuu @justinfagnani any update on this? I'd be interested in this as well!

**Comment by Noleli** [comment]
I’m quite interested in this, especially for assigning properties to slotted children.

What’s the current preferred way to do it? I’m using `@queryAssignedElements` in an `update()` callback, but a declarative way to do it in a template would be preferable.
